Objective-C/CBLProperties.mm

```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
+    NSMutableDictionary* result = [properties mutableCopy];
+    for (NSString* key in properties) {
+        id oldValue = _properties ? _properties[key] : nil;
+        id nuValue = properties[key];
+        result[key] = [self convertValue: nuValue oldValue: oldValue forKey: key];
+    }
+    return result;
+}
+
+
+- (id) convertValue: (id)value oldValue: (nullable id)oldValue forKey: (NSString*)key {
+    if ([value isKindOfClass: [NSDictionary class]]) {
+        CBLSubdocument* subdoc = nil;
+        // If the old value is a subdocument, use the old value:
+        if ([oldValue isKindOfClass: [CBLSubdocument class]]) {
+            subdoc = (CBLSubdocument*)oldValue;
+            subdoc.properties = value;
+        } else
+            subdoc = [self subdocumentWithRoot: nullptr withProperties: value forKey: key];
+        return subdoc;
+    } else if ([value isKindOfClass: [NSArray class]]) {
+        NSArray* oldArray = $castIf(NSArray, oldValue);
+        NSArray* nuArray = (NSArray*)value;
+        NSMutableArray* result = [NSMutableArray arrayWithCapacity: nuArray.count];
+        for (int i = 0; i < nuArray.count; i++) {
+            id oldValue = oldArray && i < oldArray.count ? oldArray[i] : nil;
+            id nuValue = nuArray[i];
+            result[i] = [self convertValue: nuValue oldValue: oldValue forKey: key];
+        }
+        
+        // Invalidate:
+        if (oldArray && oldArray.count > nuArray.count) {
+            for (int i = (int)nuArray.count; i < oldArray.count; i++) {
+                CBLSubdocument* oldSubdoc = $castIf(CBLSubdocument, oldArray[i]);
+                [oldSubdoc invalidate];
+            }
+        } else {
+            CBLSubdocument* oldSubdoc = $castIf(CBLSubdocument, oldValue);
+            [oldSubdoc invalidate];
+        }
+        return result;
+    } else {
+        CBLSubdocument* nuSubdoc = $castIf(CBLSubdocument, value);
+        [self attachSubdocument: nuSubdoc forKey: key];
+        
+        if ([oldValue isKindOfClass: [CBLSubdocument class]])
+            [oldValue invalidate];
+    }
+    return value;
+}
+
+
+- (NSDictionary*) currentPropertiesAsJson: (BOOL)asJson {
+    NSMutableDictionary* result;
+    NSMutableSet *changedNames = [_changedNames mutableCopy];
+    if (_root) {
+        int count = FLDict_Count(_root);
+        result = [NSMutableDictionary dictionaryWithCapacity: count];
+        if (count > 0) {
+            FLDictIterator iter;
+            FLDictIterator_Begin(_root, &iter);
+            do {
+                NSString* key = [self fleeceValueToKeyString: FLDictIterator_GetKey(&iter)];
+                id value = _properties[key];
+                if (value) {
+                    if (asJson)
+                        value = [self jsonValueForValue: value];
+                    [changedNames removeObject: key];
+                } else {
+                    value = [self fleeceValueToObject: FLDictIterator_GetValue(&iter)
```

Comment:
This doesn't work for the case of a property that's been deleted (but not yet saved). In this case the property will have a value in `_root`, but not in `_properties`. (And the key will be in `_changedNames`.)
I think the right algorithm to use is
1. Set the result to a mutable copy of `_properties`
2. Iterate through `_root`.
   1. If the key is not in `_changedKeys`, and there's no value for the key in `_properties`, decode the Fleece value and add it to the result.